automata/fa/dfa.py

```diff
@@ -710,17 +710,159 @@ def maximum_word_length(self):
             return float('inf')
 
     @classmethod
-    def from_finite_language(cls, language, input_symbols):
+    def contains_substring(cls, input_symbols, substring):
```

Comment:
@Tagl Can we rename this method to `from_substring`? That would map better nicely to the Python convention used in this library (e.g. `from_dfa`, `from_nfa`, etc.)

Reply:
I can do that and with that naming change I would like to propose an additional optional parameter which defines whether we seek containment or avoidance.
Construction remains the same but final states will be different. This allows the user to avoid the overhead of using the complement function after construction.
My main use case for example has a lot to do with containment/avoidance of subsequences.

Reply:
@Tagl I think that parameter would be great. We can convert it to an internal helper if later on we feel like we want different user-facing function names.

Reply:
> On a related note, do you feel like having a Match object would be useful to you? Something that takes in an FSM (DFA or NFA) and uses it to search for all matches in a given input string.

No, not particularly useful for me so far.

Reply:
@eliotwrobson @caleb531 I just pushed an optimization of this function. I was able to pass in strings of length `>100000` after the optimization with a much faster runtime than before.

Example repeated pattern `abcd`
| size of string | runtime in seconds |
| ----:| ----------------------:|
| 4 | 0.00027 |
| 40 | 0.00023 |
| 400 | 0.00234 |
| 4000 | 0.02422 |
| 40000 | 0.19565 |
| 400000 | 1.93628 |